staking.md

```diff
@@ -0,0 +1,285 @@
+# Staking
+
+
+#### **Constants:**
+
+- **`MINIMUM_STAKE`**: The minimum amount of tokens required to be eligible for leader selection in any given epoch.
+
+#### **Participant's Balances:**
+
+- **Total Balance (TB)**: The total number of tokens a participant has deposited in the staking contract via the `stake()` function.
+- **Locked Balance (LB)**: The max of staked amounts over the **current staking cycle**, i.e., epochs `N` to `N + C`, where `C` is the number of epochs in the staking cycle (in our case, `C = 3`).
+- **Available Balance (AB)**: The portion of the Total Balance that is not locked in the current staking cycle.
+  - **AB = TB - LB**
+- **Staking Schedule (orderdlist ({epoch, amount}))**: A mapping of epochs to the amount of tokens the participant has committed to stake in each of those epochs and onwards if not changed.
+
+Automatic Restaking: Unless adjusted, the stake for an epoch automatically continues into future cycles.
+
+---
+
+#### **API Functions**
+
+1. **`deposit()`** (Payable Endpoint)
+   - **Description**: Allows a participant to deposit tokens into the staking contract, increasing their Total Balance (TB) and Available Balance (AB).
+   - **Parameters**: None (the amount is determined by the value of tokens sent with the transaction).
+   - **When Callable**: At any time.
+   - **Behavior**:
+     - Increases TB by the amount of tokens sent.
+     - Increases AB by the same amount.
+
+2. **`withdraw(amount)`**
+   - **Description**: Allows a participant to withdraw tokens from their Available Balance (AB).
+   - **Parameters**:
+     - `amount`: The amount of tokens to withdraw.
+   - **When Callable**: At any time.
+   - **Constraints**:
+     - `amount` must be less than or equal to AB.
+     - Cannot withdraw tokens locked in the current staking cycle (i.e., tokens committed in SS for epochs `N` to `N + C`).
+
+3. **`adjustStake(amount:integer)`**
+   - **Description**: Adjusts the amount of tokens a participant wishes to stake for the **next staking cycle**.
+   - **Parameters**:
+     - `amount`: The amount to adjust (positive to increase stake, negative to decrease stake).
+   - **When Callable**: Only during the **Staking Epoch**.
```

Comment:
I could use clarification here, isn't it always _some_ epoch that you could commit a stake in?

Reply:
Yes. Every epoch is a staking epoch for some staking cycle, so you can always do this, but it only affects the cycle for which you are in the staking epoch. Ill try to make it clearer.